### ISSUE TITLE
Zero counts in raw count matrix for certain CBs

### ISSUE THREAD

**Issue by ms-gx** [alexdobin/STAR#1632]
I see a case where the raw count matrix (path `Solo.out/Gene/raw`) contains a lot of CBs with only zeros for all genes. This is for a particular analysis in my case, but I assume this is a general thing?

I assume that is because the CB was reliably detected but then there were cases such as gene multi mapping where the counts got zero but the CB is reported anyway? (If this is the case, wouldn't addressing this be a good way to decrease output size?)

EDIT: I do NOT use `--soloMultiMappers`. So it is the default (`unique`).

**Comment by alexdobin** [comment]
The UniqueAndMult-EM.mtx will contain some genes with 0 counts. These genes multi-gene reads mapping to them, but the EM assigns these reads to other genes.

**Comment by ms-gx** [comment]
Sorry, my question was not stated correctly and I corrected it. I don't know what I was thinking. Now it should be reflecting what I see.

Where do I see `UniqueAndMult-EM.mtx`?

And another question:
`Solo.out/Gene/raw` is the same thing as the raw count matrix in Cell Ranger, right?

**Comment by alexdobin** [comment]
The unique-gene output (matrix.mtx) should not contain any lines with 0 counts (i.e. field 3 = 0). What parameters are you using?

> I guess UniqueAndMult-EM.mtx is there only if I do --soloMultiMappers EM?

Correct!

> And another question, just to be sure:
Solo.out/Gene/raw is in principle the same thing as the raw count matrix for Cell Ranger, right (Ignoring some minor differences of the algorithm but in general)?

Correct!

**Comment by ms-gx** [comment]
OK, I investigated a bit more.

What happened was the following:
I read the raw matrix generated by STARsolo via `Seurat::Read10X` and then I took the sum per cell barcode to check something. This is where I got CBs with zero counts which I reported above.

Now the thing is that your matrix.mtx does NOT contain zeros (meaning zeros in third field). But the barcodes.tsv contains barcodes which do not occur in in matrix.mtx. The `Read10x` function was then just importing the barcodes with zero counts and this is what I observed.

I this expected behaviour?

My STARsolo run:
`$star_path --runThreadN 63 --genomeDir $ref_annotation_path -- readFilesIn MY_FILES --soloType CB_UMI_Simple --outFileNamePrefix "$workdir_path"MAPPING_"$id"_ --soloUMIlen 12 --soloCBwhitelist <(zcat cellranger/barcodes/3M-february-2018.txt.gz) --outSAMattributes CR CY UR UY CB UB cN --outSAMtype BAM SortedByCoordinate --soloCBmatchWLtype 1MM multi Nbase pseudocounts --soloUMIfiltering MultiGeneUMI_CR --soloUMIdedup 1MM_CR --clipAdapterType CellRanger4 --outFilterScoreMin 30 --soloCellFilter EmptyDrops_CR --limitBAMsortRAM $sort_memory --outSAMunmapped Within --soloFeatures Gene GeneFull SJ > "$workdir_path"MAPPING_"$id".mainlog.txt`
